shared/dataflow/codeql/dataflow/internal/DataFlowImpl.qll

```diff
@@ -1514,34 +1587,58 @@ module MakeImpl<InputSig Lang> {
+          fwdFlow(ret, state, cc, summaryCtx, argT, argAp, t, ap, apa)
+        }
+
         pragma[nomagic]
         private predicate fwdFlowOutCand(
-          DataFlowCall call, DataFlowCallable inner, NodeEx out, FlowState state,
-          ParamNodeOption summaryCtx, TypOption argT, ApOption argAp, Typ t, Ap ap, ApApprox apa
+          DataFlowCall call, RetNodeEx ret, CcNoCall innercc, DataFlowCallable inner, NodeEx out,
+          ApApprox apa, boolean allowsFieldFlow
         ) {
-          exists(RetNodeEx ret, boolean allowsFieldFlow, CcNoCall innercc |
-            fwdFlow(ret, state, innercc, summaryCtx, argT, argAp, t, ap, apa) and
-            inner = ret.getEnclosingCallable() and
-            (
-              inner = viableImplCallContextReducedReverseInlineLate(call, innercc) and
-              flowOutOfCallApaInlineLate(call, inner, ret, out, allowsFieldFlow, apa)
-              or
-              flowOutOfCallApaNotCallContextReduced(call, inner, ret, out, allowsFieldFlow, apa,
-                innercc)
-            )
-          |
-            if allowsFieldFlow = false then ap instanceof ApNil else any()
+          fwdFlowIntoRet(ret, _, innercc, _, _, _, _, _, apa) and
+          inner = ret.getEnclosingCallable() and
+          (
+            inner = viableImplCallContextReducedReverseInlineLate(call, innercc) and
+            flowOutOfCallApaInlineLate(call, inner, ret, out, allowsFieldFlow, apa)
+            or
+            flowOutOfCallApaNotCallContextReduced(call, inner, ret, out, allowsFieldFlow, apa,
+              innercc)
           )
         }
 
         pragma[nomagic]
+        private predicate fwdFlowOutValidEdge(
+          DataFlowCall call, RetNodeEx ret, CcNoCall innercc, DataFlowCallable inner, NodeEx out,
+          CcNoCall outercc, ApApprox apa, boolean allowsFieldFlow
+        ) {
+          fwdFlowOutCand(call, ret, innercc, inner, out, apa, allowsFieldFlow) and
+          FwdTypeFlow::typeFlowValidEdgeOut(call, inner) and
+          outercc = getCallContextReturn(inner, call)
+        }
+
+        pragma[inline]
         private predicate fwdFlowOut(
           DataFlowCall call, DataFlowCallable inner, NodeEx out, FlowState state, CcNoCall outercc,
           ParamNodeOption summaryCtx, TypOption argT, ApOption argAp, Typ t, Ap ap, ApApprox apa
         ) {
-          fwdFlowOutCand(call, inner, out, state, summaryCtx, argT, argAp, t, ap, apa) and
-          FwdTypeFlow::typeFlowValidEdgeOut(call, inner) and
-          outercc = getCallContextReturn(inner, call)
+          exists(RetNodeEx ret, CcNoCall innercc, boolean allowsFieldFlow |
+            fwdFlowIntoRet(ret, state, innercc, summaryCtx, argT, argAp, t, ap, apa) and
+            fwdFlowOutValidEdge(call, ret, innercc, inner, out, outercc, apa, allowsFieldFlow) and
+            if allowsFieldFlow = false then ap instanceof ApNil else any()
+          )
+        }
+
+        pragma[nomagic]
+        private predicate fwdFlowOut(
```

Comment:
The introduction of this projection seems superfluous. It's only used once in `fwdFlow0` and there its occurrence is as a trivial disjunct, so doing the projection there instead as we did before will be exactly the same work, except avoiding the additional materialisation.

Reply:
Oh, wait, the wider version is now inlined. Hmm, but that still shouldn't matter, right? Dropping this predicate will just postpone the projection and dedup to `fwdFlow0` where the cost will be the same, but will save this materialisation.

Reply:
I think it should be OK to inline; only reason I `nomagic`ed it is because it is sometimes convenient to be able to spot in the timing logs whether it is, say, flow out that is expensive. But I'll inline it.